src/middlewares/createInsightsMiddleware.ts

```diff
@@ -38,26 +38,35 @@ export const createInsightsMiddleware: CreateInsightsMiddleware = props => {
     _insightsClient === null ? (noop as InsightsClient) : _insightsClient;
 
   return ({ instantSearchInstance }) => {
-    insightsClient('_get', '_hasCredentials', (hasCredentials: boolean) => {
-      if (!hasCredentials) {
-        const [appId, apiKey] = getAppIdAndApiKey(instantSearchInstance.client);
-        insightsClient('_get', '_userToken', (userToken: string) => {
-          warning(
-            !userToken,
-            `You set userToken before \`createInsightsMiddleware()\` and it is ignored.
-Please set the token after the \`createInsightsMiddleware()\` call.
-
-createInsightsMiddleware({ /* ... */ });
-
-insightsClient('setUserToken', 'your-user-token');
-// or
-aa('setUserToken', 'your-user-token');
-            `
-          );
-        });
-        insightsClient('init', { appId, apiKey });
-      }
+    const [appId, apiKey] = getAppIdAndApiKey(instantSearchInstance.client);
+    let queuedUserToken: string | undefined = undefined;
+    let userTokenBeforeInit: string | undefined = undefined;
+    if (Array.isArray((insightsClient as any).queue)) {
+      // Context: The umd build of search-insights is asynchronously loaded by the snippet.
+      //
+      // When user calls `aa('setUserToken', 'my-user-token')` before `search-insights` is loaded,
+      // ['setUserToken', 'my-user-token'] gets stored in `aa.queue`.
+      // Whenever `search-insights` is finally loaded, it will process the queue.
+      //
+      // But here's the reason why we handle it here:
+      // At this point, even though `search-insights` is not loaded yet,
+      // we still want to read the token from the queue.
+      // Otherwise, the first search call will be fired without the token.
+      (insightsClient as any).queue.forEach(([method, firstArgument]) => {
```

Comment:
Why using `forEach` if we override the variable every time? Using `find` would make more sense:

```ts
find(insightsClient.queue.slice().reverse(), /* ... */)
```

- `slice` to not mutate the queue
- `reverse` to find the latest occurrence

Reply:
https://github.com/algolia/instantsearch.js/pull/4575/commits/fa7ea26d1c4c4d2a49960cd89a731d50a6fea9f2